source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -145,6 +148,18 @@ OptRef<const LocalRateLimiterImpl::LocalDescriptorImpl> LocalRateLimiterImpl::de
 
 bool LocalRateLimiterImpl::requestAllowed(
     absl::Span<const RateLimit::LocalDescriptor> request_descriptors) const {
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.http_local_ratelimit_match_all_descriptors")) {
+    bool limit = requestAllowedHelper(tokens_);
+    if (!descriptors_.empty() && !request_descriptors.empty()) {
+      for (const auto& request_descriptor : request_descriptors) {
+        if (auto it = descriptors_.find(request_descriptor); it != descriptors_.end()) {
+          limit &= requestAllowedHelper(*it->token_state_);
+        }
+      }
+    }
+    return limit;
```

Comment:
By this way, if a request is limited by a descriptor, it will still comsume token from the remaining matched descriptors.

For example, if client A is allowed to send 3 requests per second. Then we set a gloabl tokens with limit of 10 requests per second. Then fouth request of client A shouldn't comsume the token of global tokens. But in this implementation, the tokens still be comsumed.

Reply:
Great question, i will fix it.

Reply:
@wbpcode i have tried to fix this, PTAL, thanks. Now i first check the remaining tokens of matched descriptors instead of trying to consuming tokens, if one of these tokens is 0, it just return false and will not consume any tokens of matched descriptors. Otherwise, it will try to consume 1 token from each matched descriptor and always return true regardless of whether tokens are all consumed by other threads.